packages/blaze-tx/src/tx.ts

```diff
@@ -1241,167 +1243,72 @@ export class TxBuilder {
    * Prepares the collateral for the transaction by selecting suitable UTXOs.
    * Throws an error if suitable collateral cannot be found or if some inputs cannot be resolved.
    */
-  private prepareCollateral() {
+  public prepareCollateral(
+    { useCoinSelection }: { useCoinSelection: boolean } = {
+      useCoinSelection: true,
+    },
+  ) {
     // Retrieve provided collateral inputs
     const providedCollateral = [...this.collateralUtxos.values()].sort(
       (a, b) =>
         a.output().amount().coin() < b.output().amount().coin() ? -1 : 1,
     );
-    // Retrieve inputs from the transaction body and available UTXOs within scope.
-    const inputs = [...this.body.inputs().values()];
-    const scope = [...this.utxoScope.values()];
-    // Initialize variables to track the best UTXO for collateral and its ranking.
-    let [best, rank]: [TransactionUnspentOutput[] | undefined, number] = [
-      undefined,
-      99,
-    ];
-    // if there are provided collateral UTxOs, use them first
-    if (providedCollateral.length > 0) {
-      for (const utxo of providedCollateral) {
-        const coinAmount = this.getUtxoEffectiveCoin(utxo);
-        if (
-          coinAmount >= 5_000_000 &&
-          utxo.output().address().getProps().paymentPart?.type ==
-            CredentialType.KeyHash
-        ) {
-          const ranking = value.assetTypes(utxo.output().amount());
-          if (ranking < rank) {
-            rank = ranking;
-            best = [utxo];
-          }
-        }
-      }
+
+    const totalValue = value.sum(
+      providedCollateral.map((pc) => pc.output().amount()),
+    );
+
+    const requiredLovelace = BigInt(
+      Math.ceil(Number(this.fee) * (this.params.collateralPercentage / 100)),
+    );
+
+    if (!useCoinSelection) {
+      const tis = CborSet.fromCore([], TransactionInput.fromCore);
+      tis.setValues(providedCollateral.map((pc) => pc.input()));
+      this.body.setCollateral(tis);
+      this.body.setTotalCollateral(requiredLovelace);
+      this.body.setCollateralReturn(
+        new TransactionOutput(
+          this.collateralChangeAddress ?? this.changeAddress!,
+          value.merge(totalValue, value.negatives(new Value(requiredLovelace))),
+        ),
+      );
     } else {
-      // If no provided collateral inputs, iterate over the inputs to find the best candidate.
-      for (const input of inputs) {
-        const utxo = scope.find(
-          (x) =>
-            x.input().transactionId() === input.transactionId() &&
-            x.input().index() === input.index(),
+      let remainingLovelace = BigInt(requiredLovelace);
```

Comment:
Should this subtract off the manually set collateral inputs we start with?

Reply:
I don't think so, because if we're using coinSelection, then we shouldn't be providing our own collateral inputs.

Reply:
Hmm; I think the behavior I would expect is: if I set a collateral input, or add inputs to the transaction, they will always be included; but with coinselection, it will select *additional* inputs to cover the requirements.

Reply:
That makes sense, I've updated it to always include them in the total input value.